src/datasets/utils/streaming_download_manager.py

```diff
@@ -74,8 +88,13 @@ def xopen(file, mode="r", *args, **kwargs):
     """
     if fsspec.get_fs_token_paths(file)[0].protocol == "https":
         kwargs["headers"] = get_authentication_headers_for_url(file, use_auth_token=kwargs.pop("use_auth_token", None))
-    file_obj = fsspec.open(file, mode=mode, *args, **kwargs).open()
-    _add_retries_to_file_obj_read_method(file_obj)
+    compression = fsspec.core.get_compression(file, "infer")
+    if not compression or compression in ["gzip", "zip"]:
+        file_obj = fsspec.open(file, mode=mode, *args, **kwargs).open()
+        file_obj = _add_retries_to_file_obj_read_method(file_obj)
+    else:
+        file_obj = fsspec.open(file, mode=mode, compression=compression, *args, **kwargs)
+        file_obj = _add_retries_to_fsspec_open_file(file_obj)
```

Comment:
`xopen` is an extension of `open` to make it work with remote files.

Here you change its behavior for compressed files: you automatically uncompress them. Therefore if you try to open a compressed file and then use `gzip` (or any other compressing tool) to uncompress it, it won't work since it's already uncompressed.

I think we should revert this change, and explicitly use some tool in the dataset scripts to uncompress the files as we do in standard python. Otherwise we may end up with code that works in streaming mode but not in standard mode and vice-versa.

Let me know what you think @albertvillanova

Reply:
No, `fsspec.open` (even if passed the `compression` parameter) does not uncompress the file immediately: it returns an `OpenFile` instance, which will return a file-object wrapped with a decompressor instance (when called within a context manager), which will decompress on the fly... ;)

Reply:
And yes, at the end, the result (after having called `dl_manager.download_and_extract` will be an uncompressed file, either streaming or not. That is the objective! 😉

Reply:
The issue is: how do you make that `StreamingDownloadManager.extract()` passes the parameter `compression=compression` to `fsspec.open(urlpath, compression=compression)` if they can communicate only through the parameter `urlpath`?

Because of this, I always pass `compression="infer"`, which assumes that all datasets scripts have called `.extract` (or `.download_and_extract`) before calling `fsspec.open`. This assumption is sensible and will work for all dataset scripts, except for oscar (as you told me yesterday), because you changed oscar with a call: `gzip.open(open())`.